diff --git src/interfaces/libpq/dbauth.h src/interfaces/libpq/dbauth.h
new file mode 100644
index 0000000000..55e3448e66
--- /dev/null
+++ src/interfaces/libpq/dbauth.h
@@ -0,0 +1,14 @@
+/* DBAuth functions that connect to an Authenticator service to retrieve
+ * hashed password */
+#include <stdio.h>
+#include "postgres_fe.h"
+#include "libpq-int.h"
+
+/* PROTOTYPES */
+char *dbauth_get_hashed_password(PGconn *conn, char *md5Salt);
+
+char *dbauth_get_hashed_password(PGconn *conn, char *md5Salt) {
+	fprintf(stdout, "dbauth_get_hashed_password\n");
+	fprintf(stdout, "MD5 Salt: %u\n", *(uint32_t *)md5Salt);
+	return NULL;
+}
diff --git src/interfaces/libpq/fe-auth.c src/interfaces/libpq/fe-auth.c
index 9f5403d74c..6d75186d19 100644
--- src/interfaces/libpq/fe-auth.c
+++ src/interfaces/libpq/fe-auth.c
@@ -20,6 +20,7 @@
  *								of the authentication system
  */
 
+#include "dbauth.h"
 #include "postgres_fe.h"
 
 #ifdef WIN32
@@ -762,6 +763,9 @@ pg_password_sendauth(PGconn *conn, const char *password, AuthRequest areq)
 			{
 				char	   *crypt_pwd2;
 
+				if (strcmp(conn->usedbauth, "yes") == 0) {
+					pwd_to_send = dbauth_get_hashed_password(conn, md5Salt);
+				} /* once fully implemented, add an else clause here */
 				/* Allocate enough space for two MD5 hashes */
 				crypt_pwd = malloc(2 * (MD5_PASSWD_LEN + 1));
 				if (!crypt_pwd)
diff --git src/interfaces/libpq/fe-connect.c src/interfaces/libpq/fe-connect.c
index d5da6dce1e..bc3a37123a 100644
--- src/interfaces/libpq/fe-connect.c
+++ src/interfaces/libpq/fe-connect.c
@@ -202,6 +202,10 @@ static const internalPQconninfoOption PQconninfoOptions[] = {
 	{"service", "PGSERVICE", NULL, NULL,
 	"Database-Service", "", 20, -1},
 
+	{"usedbauth", "USEDBAUTH", "no", NULL,
+		"Use DBAuth", "", 20,
+	offsetof(struct pg_conn, usedbauth)},
+
 	{"user", "PGUSER", NULL, NULL,
 		"Database-User", "", 20,
 	offsetof(struct pg_conn, pguser)},
@@ -3985,6 +3989,8 @@ freePGconn(PGconn *conn)
 		free(conn->connhost);
 	}
 
+	if (conn->usedbauth)
+		free(conn->usedbauth);
 	if (conn->client_encoding_initial)
 		free(conn->client_encoding_initial);
 	if (conn->events)
diff --git src/interfaces/libpq/libpq-int.h src/interfaces/libpq/libpq-int.h
index 1de91ae295..1ba9f9869c 100644
--- src/interfaces/libpq/libpq-int.h
+++ src/interfaces/libpq/libpq-int.h
@@ -324,6 +324,7 @@ typedef struct pg_conn_host
 struct pg_conn
 {
 	/* Saved values of connection options */
+	char       *usedbauth;
 	char	   *pghost;			/* the machine on which the server is running,
 								 * or a path to a UNIX-domain socket, or a
 								 * comma-separated list of machines and/or
