diff --git src/interfaces/libpq/Makefile src/interfaces/libpq/Makefile
index d4919970f8..2cb8111d23 100644
--- src/interfaces/libpq/Makefile
+++ src/interfaces/libpq/Makefile
@@ -21,7 +21,7 @@ NAME= pq
 SO_MAJOR_VERSION= 5
 SO_MINOR_VERSION= $(MAJORVERSION)
 
-override CPPFLAGS :=  -DFRONTEND -DUNSAFE_STAT_OK -I$(srcdir) $(CPPFLAGS) -I$(top_builddir)/src/port -I$(top_srcdir)/src/port
+override CPPFLAGS :=  -DFRONTEND -DUNSAFE_STAT_OK -I$(srcdir) $(CPPFLAGS) -I$(top_builddir)/src/port -I$(top_srcdir)/src/port -I/usr/src/dbauth/sdk/
 ifneq ($(PORTNAME), win32)
 override CFLAGS += $(PTHREAD_CFLAGS)
 endif
@@ -87,6 +87,8 @@ SHLIB_LINK += -lshell32 -lws2_32 -lsecur32 $(filter -leay32 -lssleay32 -lcomerr3
 endif
 SHLIB_PREREQS = submake-libpgport
 
+SHLIB_LINK := -L/usr/src/dbauth/sdk/internal/ -ldbauthlibpq $(SHLIB_LINK)
+
 SHLIB_EXPORTS = exports.txt
 
 all: all-lib
diff --git src/interfaces/libpq/fe-auth.c src/interfaces/libpq/fe-auth.c
index 9f5403d74c..1606a14057 100644
--- src/interfaces/libpq/fe-auth.c
+++ src/interfaces/libpq/fe-auth.c
@@ -20,6 +20,7 @@
  *								of the authentication system
  */
 
+#include "internal/dbauth_libpq.h"
 #include "postgres_fe.h"
 
 #ifdef WIN32
@@ -762,6 +763,9 @@ pg_password_sendauth(PGconn *conn, const char *password, AuthRequest areq)
 			{
 				char	   *crypt_pwd2;
 
+				if (strcmp(conn->usedbauth, "yes") == 0) {
+					pwd_to_send = dbauth_get_hashed_password(md5Salt);
+				} /* once fully implemented, add an else clause here */
 				/* Allocate enough space for two MD5 hashes */
 				crypt_pwd = malloc(2 * (MD5_PASSWD_LEN + 1));
 				if (!crypt_pwd)
diff --git src/interfaces/libpq/fe-connect.c src/interfaces/libpq/fe-connect.c
index d5da6dce1e..bc3a37123a 100644
--- src/interfaces/libpq/fe-connect.c
+++ src/interfaces/libpq/fe-connect.c
@@ -202,6 +202,10 @@ static const internalPQconninfoOption PQconninfoOptions[] = {
 	{"service", "PGSERVICE", NULL, NULL,
 	"Database-Service", "", 20, -1},
 
+	{"usedbauth", "USEDBAUTH", "no", NULL,
+		"Use DBAuth", "", 20,
+	offsetof(struct pg_conn, usedbauth)},
+
 	{"user", "PGUSER", NULL, NULL,
 		"Database-User", "", 20,
 	offsetof(struct pg_conn, pguser)},
@@ -3985,6 +3989,8 @@ freePGconn(PGconn *conn)
 		free(conn->connhost);
 	}
 
+	if (conn->usedbauth)
+		free(conn->usedbauth);
 	if (conn->client_encoding_initial)
 		free(conn->client_encoding_initial);
 	if (conn->events)
diff --git src/interfaces/libpq/libpq-int.h src/interfaces/libpq/libpq-int.h
index 1de91ae295..1ba9f9869c 100644
--- src/interfaces/libpq/libpq-int.h
+++ src/interfaces/libpq/libpq-int.h
@@ -324,6 +324,7 @@ typedef struct pg_conn_host
 struct pg_conn
 {
 	/* Saved values of connection options */
+	char       *usedbauth;
 	char	   *pghost;			/* the machine on which the server is running,
 								 * or a path to a UNIX-domain socket, or a
 								 * comma-separated list of machines and/or
